3196 profile media file

hyi opened the pull request:
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [ ] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]
This PR addressed the issue #3196. In addition to UI update/fix, the deleted files are now deleted from the system as they should. Specifically, when a profile picture or a CV is deleted, the picture and CV file is deleted from the system media file directory. A management command is also added to clean up the left over deleted files from profile media file directory that are not part of HydroShare user profiles.

hydrocheck commented:
| METRIC | VALUE |
|:-----:|:-----:|
| NEW VULNERABILITIES | 0 |
| NEW CODE SMELLS | 2 |
| NEW SECURITY HOTSPOTS | 0 |
| NEW BUGS | 0 |
| NEW DUPLICATED LINES DENSITY | 0.0 |

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4254](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4254)

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
 pre_save.connect(force_unique_emails, sender=User)
+
+
+@receiver(models.signals.pre_save, sender=UserProfile)
```
So this triggers anytime the UserProfile is updated?

hyi replied:
Yes, it is triggered anytime when user profile changes are saved. I cannot think of any more efficient way to achieve automatic deletion when a user profile CV or profile picture is updated without plugging in this trigger. I don't think user profiles are updated that often, so I don't think we need to worry about the performance, but let me know if you have better/more efficient ideas to achieve this automatic profile file cleanup.

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
+    except UserProfile.DoesNotExist:
+        return
+
+    new_file_cv = instance.cv
```
Ok, so instance is the updated model and you're reading the old model from the db and comparing to see if it changed.  I think updating variable names and comments could help clarify what is going on.  If instance were renamed to updated_profile then I'd say removing assigning new_file_cv variable and just access it in the if statement below.

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
+
+
+@receiver(models.signals.pre_save, sender=UserProfile)
+def auto_delete_file_on_change(sender, instance, **kwargs):
```
Update instance to something like updated_profile

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
+    Deletes old file from filesystem when corresponding UserProfile object is updated with new file.
+    """
+    if not instance.pk:
+        return
```
A comment here explaining that if there is no pk, then it's a new profile.

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
+            old_file_pic = up.picture
+            if not old_file_cv and not old_file_pic:
+                return
+        else:
```
This else block will never be executed, get will thrown an exception (the one you're catching) if the profile doesn't exist.

sblack-usu commented on theme/models.py:
```diff
@@ -324,4 +325,36 @@ def force_unique_emails(sender, instance, **kwargs):
+        else:
+            return
+    except UserProfile.DoesNotExist:
+        return
```
This is a pretty big deal if we get here, I'd like it logged as a warning.  I don't think this will ever happen because of the instance.pk check earlier.

sblack-usu reviewed (commented):
Have you tried to tackle the deployment issues with swapping workers yet?

hyi commented:
@sblack-usu I thought the deployment issues with swapping workers are all isolated within Jenkins job script, so there is no need for any code update in the repo. I'm counting on you and/or @cbcunc to update Jenkins job to make sure the old media file backup will not override the latest media files from the active worker.

hyi commented:
@sblack-usu Have addressed all your code review comments with one exception. Specifically, ```instance``` cannot be renamed since it is part of Django model pre_save signal providing_args signature, so renaming it to any other name will result in an exception. I did add a comment to explain it and handled the name better. See if everything looks good now.
